### PR TITLE
docs: Enable docs from remote plugin

### PR DESCRIPTION
**Pull request by zchsh**
🎟️ [Asana project](https://app.asana.com/0/1199972659301420/1199913030172635)
🔍 [Preview](https://packer-git-fork-zchsh-zsremote-builder-content.hashicorp.vercel.app/docs)

This PR pulls in new `DocsPage` and `DocsSidenav` which have been [revised](https://github.com/hashicorp/react-components/pull/154) to implement [MKTG-032](https://docs.google.com/document/d/1kYvbyd6njHFSscoE1dtDNHQ3U8IzaMdcjOS0jg87rHg). It builds on this work to enable loading of documentation content from remote repositories.

Specifically, it leverages this functionality to allow fetching of docs from remote `packer-plugin` repositories, which are automatically included in the appropriate `Builders`, `Data Sources`, `Post-Processors`, or `Provisioners` sections of the [documentation website](https://www.packer.io/docs).

### CODE REVIEW

**Comment by SwampDragons** [comment]
This seems really slick!  Can't wait to test it out.

File: website/components/remote-plugin-docs/server.js
Comment (zchsh) [inline website/components/remote-plugin-docs/server.js:8]: Currently, Packer remote plugin docs are being treated as an individual use case.

To meet this use case, we've built on the [revised DocsPage and DocsSidenav work we've done in implementing MKTG-032](https://github.com/hashicorp/react-components/pull/154).

This `server.js` file composes functions used in the revised DocsPage `server.js` utility, and does additional work, namely fetching and merging the remote plugin docs NavNodes.

File: website/data/docs-nav-data.json
Comment (zchsh) [inline website/data/docs-nav-data.json:1]: As part of this work, I've converted `navigation.js` files to `nav-data.json` files per [MKTG-032](https://docs.google.com/document/u/2/d/1FPyWid2NmNI9E_6lH2UIMuEPqjqDCGe3F8wraPRQHQ8).

File: website/data/docs-remote-plugins.json
Comment (zchsh) [inline website/data/docs-remote-plugins.json:11]: I've left some entries here for demo purposes. I'll remove these once we're happy with this approach and feel this PR is ready to be merged. Then, they can be added back into this file in the PRs that move the plugin source out of `packer`, eg #10695 .

- [ ] @zchsh  Change this file to be an empty array before merging

File: website/package.json
Comment (zchsh) [inline website/package.json:10]: I'm using a pre-release of `react-docs-page` at the moment.

We'll aim to cut a stable release before merging this PR, though it will likely be functionally identical to the pre-release, so if time is of the essence, we may choose to merge with the pre-release to unblock other work, and then quickly follow up to upgrade to a stable release.

- [ ] @zchsh Cut stable pre-release of `@hashicorp/react-docs-page`, update in deps

File: website/pages/docs/[[...page]].jsx
Comment (zchsh) [inline website/pages/docs/[[...page]].jsx:8]: On the surface, these `[[...page]]` route files should hopefully look quite familiar.

File: website/pages/guides/[[...page]].jsx
Comment (zchsh) [inline website/pages/guides/[[...page]].jsx:16]: The `guides` and `intro` routes provided an opportunity to prove out the "core use case" (local content only) for the [MKTG-032](https://docs.google.com/document/d/1kYvbyd6njHFSscoE1dtDNHQ3U8IzaMdcjOS0jg87rHg) revisions we've made to DocsPage and DocsSidenav components.

Pages under these routes will render identically to how they do now, the only difference is they now use the updated `nav-data.json` format.

File: website/pages/style.css
Comment (zchsh) [inline website/pages/style.css:15]: This component now uses CSS modules, so a global style import is no longer needed.

File: website/components/remote-plugin-docs/utils/fetch-github-file.js
Comment (zchsh) [inline website/components/remote-plugin-docs/utils/fetch-github-file.js:3]: This utility is what lets us fetch files, such as `nav-data.json` and `content.mdx`, from remote plugin repositories via the GitHub API. I've configured the `GITHUB_API_TOKEN` for both `production` and `preview` environments in the [Vercel env variables](https://vercel.com/hashicorp/packer/settings/environment-variables).

🚨 Note that going forward, authors will need to add a valid `GITHUB_API_TOKEN` (ie, a Personal Access Token) to their `.env.local` file in the root of the repository in order for the site to build as expected.

File: website/components/remote-plugin-docs/utils/merge-remote-plugins.js
Comment (zchsh) [inline website/components/remote-plugin-docs/utils/merge-remote-plugins.js:2]: This `merge-remote-plugins` function runs the job of:
1. fetching remote plugin `nav-data` (each remote plugin may contain multiple documented components) and
2. merging each documented component into the `navData` tree.

Note that this "merging" occurs _before_ we validate the `navData`. This means that if the process of merging the remote plugin entries into the `navData` creates an invalid structure (such as duplicate paths), those issues **will** be surfaced as errors.

File: website/components/plugin-tier-label/index.js
Comment (nywilken) [reply website/components/plugin-tier-label/index.js:11]: Amazing stuff.

File: website/components/remote-plugin-docs/utils/fetch-github-file.js
Comment (nywilken) [reply website/components/remote-plugin-docs/utils/fetch-github-file.js:3]: > Note that going forward, authors will need to add a valid GITHUB_API_TOKEN (ie, a Personal Access Token) to their .env.local file in the root of the repository in order for the site to build as expected.

Is this warning exclusively for folks wanting to run the website locally? If so I think that is okay and we need to make sure it is captured in the website README. 

What happens if we can't fetch the remote location. Do we get a blank page or error?

Comment (zchsh) [reply website/components/remote-plugin-docs/utils/fetch-github-file.js:3]: @nywilken yep, this is exclusively for folks wanting to run the website locally. PR deploy previews, and of course the production website, will have access to a `GITHUB_API_TOKEN` from our Vercel env. I'll add a note to the README to this effect, that's a great call! 🙇 

Locally, what happens is the build will throw an error. In local development, the error will look something like:

CleanShot 2021-02-26 at 16 35 54

File: website/components/plugin-tier-label/index.js
Comment (nywilken) [inline website/components/plugin-tier-label/index.js:8]: Love that this is a simple map. So easy to change if needed :smile:

File: website/components/remote-plugin-docs/utils/fetch-github-file.js
Comment (nywilken) [reply website/components/remote-plugin-docs/utils/fetch-github-file.js:3]: Cool - very description error messaging that tells the user what to do next. Love it!

File: website/data/docs-remote-plugins.json
Comment (nywilken) [reply website/data/docs-remote-plugins.json:6]: Pretty soon we can replace this with an an official example https://github.com/hashicorp/packer-plugin-docker/pull/1

File: website/package.json
Comment (nywilken) [reply website/package.json:10]: No issues here. This will be released onto master and not pushed to production today so there is time. As long as we can render via preview on https://packer-git-master-hashicorp.vercel.app/ we are good to go and can follow up with an official release when ready.

File: website/pages/docs/[[...page]].jsx
Comment (nywilken) [reply website/pages/docs/[[...page]].jsx:8]: :+1:

File: website/data/docs-remote-plugins.json
Comment (zchsh) [inline website/data/docs-remote-plugins.json:1]: I've left an empty array here for now. Entries for remote plugins would be something like:

```
[
  {
    title: "Docker",
    path: "docker",
    repo: "hashicorp/packer-plugin-docker",
  }
]
```

Entries can optionally also contain `branch` and `artifactDir` values, these default to `main` and `.docs-artifacts` respectively.

File: website/components/remote-plugin-docs/utils/fetch-github-file.js
Comment (jescalan) [reply website/components/remote-plugin-docs/utils/fetch-github-file.js:3]: Do we get an error trying to view any page without a local token, or just the pages that we need the token for? Ideally it would be the former 😀

File: website/components/remote-plugin-docs/utils/fetch-github-file.js
Comment (jescalan) [inline website/components/remote-plugin-docs/utils/fetch-github-file.js:21]: you should be able to load it from extension `.graphql`, i think nextjs-scripts injects a plaintext loader at this extension by default

File: website/components/remote-plugin-docs/utils/fetch-github-file.js
Comment (jescalan) [inline website/components/remote-plugin-docs/utils/fetch-github-file.js:89]: nice, this is fantastic

File: website/components/remote-plugin-docs/utils/merge-remote-plugins.js
Comment (jescalan) [inline website/components/remote-plugin-docs/utils/merge-remote-plugins.js:12]: 🔪

```suggestion
```

File: website/data/docs-nav-data.json
Comment (jescalan) [inline website/data/docs-nav-data.json:1224]: 👏🏼

File: website/data/docs-remote-plugins.json
Comment (jescalan) [reply website/data/docs-remote-plugins.json:1]: Maybe add a commented out example of the format here just for the future?

**Review by jescalan — APPROVED** [review]
Really impressive work here, this is quite an accomplishment 👏🏼

File: website/data/docs-remote-plugins.json
Comment (zchsh) [reply website/data/docs-remote-plugins.json:1]: It's only JSON5 that has comments though, right? Will definitely add some docs to the README on this 👍

File: website/components/remote-plugin-docs/utils/fetch-github-file.js
Comment (zchsh) [reply website/components/remote-plugin-docs/utils/fetch-github-file.js]: Ah yeah, makes sense! I think I did this when developing the scripts, just so I could easily run it. But I just realized that we might also need to use this outside NextJS when indexing content, so going to leave it inline for now till we can resolve the indexing challenge 🤖

File: website/components/remote-plugin-docs/utils/fetch-github-file.js
Comment (zchsh) [reply website/components/remote-plugin-docs/utils/fetch-github-file.js:3]: The latter, unfortunately - in this specific case, all `docs` pages, since they all rely on the fetch requests to know which URLs need to be rendered in `getStaticPaths`.

As discussed in Slack, I'll look into a way we can avoid this - I think we _want_ these errors to throw in production, to alert us to missing content. So I'll target dev with NODE_ENV, and if GITHUB_API_TOKEN isn't defined, I'll see if I can add a big red box to the top of all `docs` pages if it's NOT defined. (If it is defined, but is invalid, or if GitHub is down, then in local dev folks would still see the error above).

**Review by nywilken — APPROVED** [review]
This looks great, a full end to end test was deployed to https://packer-9qi5zltix-hashicorp.vercel.app/docs/builders/docker

I see there is some additional content around the indexing, and a few comments for @jescalan.  I'll leave this open, but let me know if you're at a good place where we can merge. We can PR the follow ups. Nicely done.

**Review by sylviamoss — COMMENTED** [review]
This is looking awesome @zchsh! Thanks for all of the work to get this done. 🤗

File: website/components/remote-plugin-docs/utils/fetch-github-file.js
Comment (azr) [reply website/components/remote-plugin-docs/utils/fetch-github-file.js:3]: Adding here that Github's public REST api allows to do a few requests for free without a token: maybe it could be nice in dev mode, because a dev will go and check their files.

```shellsession
> curl https://api.github.com/repos/azr/packer-plugin-amazon/contents/main.go | jq -r .content | base64 --decode
package main

import (
	"fmt"
	"os"

	"github.com/hashicorp/packer/builder/amazon/ebs"
	"github.com/hashicorp/packer/builder/amazon/ebssurrogate"
	"github.com/hashicorp/packer/builder/amazon/ebsvolume"
	"github.com/hashicorp/packer/builder/osc/chroot"
	"github.com/hashicorp/packer/packer-plugin-sdk/plugin"
	amazonimport "github.com/hashicorp/packer/post-processor/amazon-import"
)

func main() {
	pps := plugin.NewSet()
	pps.RegisterBuilder("ebs", new(ebs.Builder))
	pps.RegisterBuilder("chroot", new(chroot.Builder))
	pps.RegisterBuilder("ebssurrogate", new(ebssurrogate.Builder))
	pps.RegisterBuilder("ebsvolume", new(ebsvolume.Builder))
	pps.RegisterPostProcessor("import", new(amazonimport.PostProcessor))
	err := pps.Run()
	if err != nil {
		fmt.Fprintln(os.Stderr, err.Error())
		os.Exit(1)
	}
}
```

File: website/components/remote-plugin-docs/utils/fetch-github-file.js
Comment (azr) [reply website/components/remote-plugin-docs/utils/fetch-github-file.js:70]: +1 !

File: website/components/remote-plugin-docs/utils/fetch-github-file.js
Comment (zchsh) [reply website/components/remote-plugin-docs/utils/fetch-github-file.js:3]: @azr I love that idea - being able to show that remote previews without a `GITHUB_API_TOKEN` would be huge. 

We also have some ideas about how this could work outside the `packer` repo - eg one is that we could allow running `npx -p @hashicorp/packer-docs-artifacts preview` within plugin repositories (or really any directory with a `docs` folder), and seeing a live, accurate preview of just that individual plugin's docs and sidebar, with live reload. This would remove the need for API requests altogether (at least just for the isolated preview use case).

For now I've taken a slightly less ideal route.... but hopefully still better than just throwing an error and requiring a `GITHUB_API_TOKEN` to preview _anything_ on the `docs` route. If there's no `GITHUB_API_TOKEN`, **and** we detect things are running locally (via our `VERCEL_ENV` var), then we'll show a nice clear message on `docs` routes rather than throwing an error:

CleanShot 2021-03-02 at 19 00 53@2x

This way, local builds won't immediately break and force the addition of a `GITHUB_API_TOKEN` when we ship this feature ☺️  CC @jescalan @nywilken as this is an update to the earlier comment

> Note: in production, and in deploy previews, we will still throw an error if we can't access remote content.

File: website/components/remote-plugin-docs/utils/merge-remote-plugins.js
Comment (zchsh) [reply website/components/remote-plugin-docs/utils/merge-remote-plugins.js]: ✅

File: website/components/remote-plugin-docs/utils/merge-remote-plugins.js
Comment (zchsh) [inline website/components/remote-plugin-docs/utils/merge-remote-plugins.js:32]: In production, we don't want to build the site with missing content. In dev, we don't want to force authors to have a `GITHUB_API_TOKEN` just to preview any other part of the site, so we just warn that there is missing content, both here in standard output and in a little preview bar in the actual site.

File: website/pages/docs/[[...page]].jsx
Comment (zchsh) [inline website/pages/docs/[[...page]].jsx:26]: This shows an alert bar in local dev previews for missing remote content ( if `GITHUB_API_TOKEN` isn't set ). This lets authors know that they're not seeing the full docs site picture, without forcing them to generate a token to use just to preview the site.

File: website/components/remote-plugin-docs/server.js
Comment (jescalan) [inline website/components/remote-plugin-docs/server.js:58]: 🚀

**Review by jescalan — APPROVED** [review]
This looks great, love the addition of the dev warning!

**Review by azr — APPROVED** [review]
This is awesome !!!

**Pull request merged**
